### ISSUE TITLE
rpmvenv fails if project is not already built and contains scripts. 

### ISSUE THREAD

**Issue by gbtimmon** [kevinconway/rpmvenv#67]
project directory:
```
.
./setup.py
./MyModule
./MyModule/VERSION
./MyModule/__init__.py
./MyModule/dostuff.py
./bin
./bin/script1
./MyModule.rpmvenv.json
./requirements.txt
```

./setup.py:
```
from setuptools import setup

if __name__ == '__main__':
    setup(
        name=package_name,
        version='1.0.0',
        packages=['MyModule'],
        include_package_data=True,
        data_files=[('MyModule', ['./VERSION'])],
        zip_safe=False,
        scripts=['./bin/script1']
    )
```

./MyModule.rpmvenv.json:
```
{
    "core": {
         "version": "1.0.0",
         "group": "Application/System",
         "name": "MyModule",
         "license": "MIT",
         "summary": "short package summary"
    },
    "blocks": {
        "desc": ["some long package description",
                 "each array element is a new line"]},
    "extensions": {
        "enabled": ["python_venv", "blocks"]
    },
    "python_venv": {
        "path": "/opt/myCompany",
        "flags": [],
        "name": "MyModule",
        "pip_flags": "--trusted-host pypi.org --trusted-host files.pythonhosted.org --index-url http://localhost:8080/simple"
    }
}
```

Broken Case:
```
(user1@localhost)(/tmp/MyModule)# rpmvenv ./MyModule.rpmvenv.json
There was an error generating the RPM.
The exit code was: 1.
The rpmbuild command was: rpmbuild -ba --define='_topdir /tmp/rpmvenvDvZXA7' /tmp/rpmvenvDvZXA7/SOURCES/package.spec.
The stderr was: + umask 022
+ cd /tmp/rpmvenvDvZXA7/BUILD
+ rm -rf '/tmp/rpmvenvDvZXA7/BUILDROOT/MyModule-1.0.0-1.x86_64/*'
+ mkdir -p /tmp/rpmvenvDvZXA7/BUILDROOT/MyModule-1.0.0-1.x86_64//opt/myCompany/MyModule
+ exit 0
+ umask 022
+ cd /tmp/rpmvenvDvZXA7/BUILD
+ '[' /tmp/rpmvenvDvZXA7/BUILDROOT/MyModule-1.0.0-1.x86_64 '!=' / ']'
+ rm -rf /tmp/rpmvenvDvZXA7/BUILDROOT/MyModule-1.0.0-1.x86_64
++ dirname /tmp/rpmvenvDvZXA7/BUILDROOT/MyModule-1.0.0-1.x86_64
+ mkdir -p /tmp/rpmvenvDvZXA7/BUILDROOT
+ mkdir /tmp/rpmvenvDvZXA7/BUILDROOT/MyModule-1.0.0-1.x86_64
+ virtualenv /tmp/rpmvenvDvZXA7/BUILDROOT/MyModule-1.0.0-1.x86_64//opt/myCompany/MyModule
+ cd /tmp/rpmvenvDvZXA7/SOURCES/Issue
+ /tmp/rpmvenvDvZXA7/BUILDROOT/MyModule-1.0.0-1.x86_64//opt/myCompany/MyModule/bin/python /tmp/rpmvenvDvZXA7/BUILDROOT/MyModule-1.0.0-1.x86_64//opt/myCompany/MyModule/bin/pip install --trusted-host pypi.org --trusted-host files.pythonhosted.org --index-url http://localhost:8080/simple -r requirements.txt
You must give at least one requirement to install (see "pip help install")
You are using pip version 9.0.1, however version 19.0.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
+ cd -
+ cd /tmp/rpmvenvDvZXA7/SOURCES/Issue
+ /tmp/rpmvenvDvZXA7/BUILDROOT/MyModule-1.0.0-1.x86_64//opt/myCompany/MyModule/bin/python setup.py install
+ cd -
+ find /tmp/rpmvenvDvZXA7/BUILDROOT/MyModule-1.0.0-1.x86_64 -name RECORD -exec rm -rf '{}' ';'
+ venvctrl-relocate --source=/tmp/rpmvenvDvZXA7/BUILDROOT/MyModule-1.0.0-1.x86_64//opt/myCompany/MyModule --destination=//opt/myCompany/MyModule
+ xargs -r strip
+ find /tmp/rpmvenvDvZXA7/BUILDROOT/MyModule-1.0.0-1.x86_64//opt/myCompany/MyModule/lib -type f -name '*.so'
+ /usr/lib/rpm/check-buildroot
error: Bad exit status from /var/tmp/rpm-tmp.T61KyS (%install)
    Bad exit status from /var/tmp/rpm-tmp.T61KyS (%install)
.
The stdout was: Executing(%prep): /bin/sh -e /var/tmp/rpm-tmp.MwyCoA
Executing(%install): /bin/sh -e /var/tmp/rpm-tmp.T61KyS
New python executable in /tmp/rpmvenvDvZXA7/BUILDROOT/MyModule-1.0.0-1.x86_64/opt/myCompany/MyModule/bin/python
Installing setuptools, pip, wheel...done.
/tmp/rpmvenvDvZXA7/BUILD
running install
running bdist_egg
running egg_info
creating MyModule.egg-info
writing MyModule.egg-info/PKG-INFO
writing top-level names to MyModule.egg-info/top_level.txt
writing dependency_links to MyModule.egg-info/dependency_links.txt
writing manifest file 'MyModule.egg-info/SOURCES.txt'
reading manifest file 'MyModule.egg-info/SOURCES.txt'
writing manifest file 'MyModule.egg-info/SOURCES.txt'
installing library code to build/bdist.linux-x86_64/egg
running install_lib
running build_py
creating build
creating build/lib
creating build/lib/MyModule
copying MyModule/__init__.py -> build/lib/MyModule
copying MyModule/dostuff.py -> build/lib/MyModule
creating build/bdist.linux-x86_64
creating build/bdist.linux-x86_64/egg
creating build/bdist.linux-x86_64/egg/MyModule
copying build/lib/MyModule/__init__.py -> build/bdist.linux-x86_64/egg/MyModule
copying build/lib/MyModule/dostuff.py -> build/bdist.linux-x86_64/egg/MyModule
byte-compiling build/bdist.linux-x86_64/egg/MyModule/__init__.py to __init__.pyc
byte-compiling build/bdist.linux-x86_64/egg/MyModule/dostuff.py to dostuff.pyc
installing package data to build/bdist.linux-x86_64/egg
running install_data
copying ./MyModule/VERSION -> build/bdist.linux-x86_64/egg/MyModule
creating build/bdist.linux-x86_64/egg/EGG-INFO
installing scripts to build/bdist.linux-x86_64/egg/EGG-INFO/scripts
running install_scripts
running build_scripts
creating build/scripts-2.7
copying and adjusting ./bin/script1 -> build/scripts-2.7
changing mode of build/scripts-2.7/script1 from 644 to 755
creating build/bdist.linux-x86_64/egg/EGG-INFO/scripts
copying build/scripts-2.7/script1 -> build/bdist.linux-x86_64/egg/EGG-INFO/scripts
changing mode of build/bdist.linux-x86_64/egg/EGG-INFO/scripts/script1 to 755
copying MyModule.egg-info/PKG-INFO -> build/bdist.linux-x86_64/egg/EGG-INFO
copying MyModule.egg-info/SOURCES.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
copying MyModule.egg-info/dependency_links.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
copying MyModule.egg-info/not-zip-safe -> build/bdist.linux-x86_64/egg/EGG-INFO
copying MyModule.egg-info/top_level.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
creating dist
creating 'dist/MyModule-1.0.0-py2.7.egg' and adding 'build/bdist.linux-x86_64/egg' to it
removing 'build/bdist.linux-x86_64/egg' (and everything under it)
Processing MyModule-1.0.0-py2.7.egg
creating /tmp/rpmvenvDvZXA7/BUILDROOT/MyModule-1.0.0-1.x86_64/opt/myCompany/MyModule/lib/python2.7/site-packages/MyModule-1.0.0-py2.7.egg
Extracting MyModule-1.0.0-py2.7.egg to /tmp/rpmvenvDvZXA7/BUILDROOT/MyModule-1.0.0-1.x86_64/opt/myCompany/MyModule/lib/python2.7/site-packages
Adding MyModule 1.0.0 to easy-install.pth file
Installing script1 script to /tmp/rpmvenvDvZXA7/BUILDROOT/MyModule-1.0.0-1.x86_64/opt/myCompany/MyModule/bin

Installed /tmp/rpmvenvDvZXA7/BUILDROOT/MyModule-1.0.0-1.x86_64/opt/myCompany/MyModule/lib/python2.7/site-packages/MyModule-1.0.0-py2.7.egg
Processing dependencies for MyModule==1.0.0
Finished processing dependencies for MyModule==1.0.0
/tmp/rpmvenvDvZXA7/BUILD
/tmp/rpmvenvDvZXA7/BUILDROOT/MyModule-1.0.0-1.x86_64/opt/myCompany/MyModule/lib/python2.7/site-packages/MyModule-1.0.0-py2.7.egg/EGG-INFO/scripts/script1:#!/tmp/rpmvenvDvZXA7/BUILDROOT/MyModule-1.0.0-1.x86_64/opt/myCompany/MyModule/bin/python
Found '/tmp/rpmvenvDvZXA7/BUILDROOT/MyModule-1.0.0-1.x86_64' in installed files; aborting


RPM build errors:
.
```

Working Case
```
(user1@localhost)(/tmp/MyModule)#./setup.py build
running build
running build_py
creating build
creating build/lib
creating build/lib/MyModule
copying MyModule/__init__.py -> build/lib/MyModule
copying MyModule/dostuff.py -> build/lib/MyModule
running egg_info
creating MyModule.egg-info
writing MyModule.egg-info/PKG-INFO
writing top-level names to MyModule.egg-info/top_level.txt
writing dependency_links to MyModule.egg-info/dependency_links.txt
writing manifest file 'MyModule.egg-info/SOURCES.txt'
reading manifest file 'MyModule.egg-info/SOURCES.txt'
writing manifest file 'MyModule.egg-info/SOURCES.txt'
running build_scripts
creating build/scripts-2.7
copying and adjusting ./bin/script1 -> build/scripts-2.7
changing mode of build/scripts-2.7/script1 from 644 to 755
(user1@localhost)(/tmp/MyModule)# rpmvenv ./MyModule.rpmvenv.json
RPM generated at ./MyModule-1.0.0-1.x86_64.rpm
```

Seems to be an issue somewhere in venvctrl relocate not getting the shebangs in the bin *unless* its already built (?). Work around is simple but shouldnt be necessary I suppose. 

This is a wonderful lib and we almost gave up on it because we couldn't figure out what was causing what appeared to be intermittent failures -- so if this is intentional design a sensible error message or revised documentation would be nice.

**Comment by kevinconway** [comment]
I think there is a combination of issues causing the failure.

For one, I think you've identified a bug in `venvctrl`. Specifically, `venvctrl` is operating under the expectation that all scripts installed in the `virtualenv` will be placed in the `bin` directory. Looking at the output you shared it would appear that binary distribution eggs retain a copy of their scripts inside the egg directory. I believe it is also copying the script to `bin` and the copy in `bin` is being rewritten appropriately based on:

> Installing script1 script to /tmp/rpmvenvDvZXA7/BUILDROOT/MyModule-1.0.0-1.x86_64/opt/myCompany/MyModule/bin

However, nothing in `venvctrl` is scanning for eggs that might also contain content which is what is causing the failure here:

> /tmp/rpmvenvDvZXA7/BUILDROOT/MyModule-1.0.0-1.x86_64/opt/myCompany/MyModule/lib/python2.7/site-packages/MyModule-1.0.0-py2.7.egg/EGG-INFO/scripts/script1:#!/tmp/rpmvenvDvZXA7/BUILDROOT/MyModule-1.0.0-1.x86_64/opt/myCompany/MyModule/bin/python

I think it would be reasonable to open an issue for `venvctrl` that links to this issue for the output and commentary. However, given time constraints and the deprecated status of the egg format, it is unlikely that I'll resolve the issue in the near term. If you, or any future reader of this, are interested in making a PR, though, I'm happy to help with reviews and getting an updated package pushed to PyPi.

A second issue that I'm not sure, yet, how to attribute is the fact that `setup.py build` and `setup.py install` appear to use different strategies when you run them. For example, your output from `setup.py build` shows that `setuptools` is creating a source distribution:

> (user1@localhost)(/tmp/MyModule)#./setup.py build
> running build
> running build_py

However, when `rpmvenv` is running `setup.py install` on the same project it is building a binary distribution egg:

> /tmp/rpmvenvDvZXA7/BUILD
> running install
> running bdist_egg
> running egg_info

It's not clear to me why these two commands would generate different output. My reading of the `install` command is that it calls `build` and then moves the resulting files to the right location so it _should_ execute the same build strategy. Source distributions don't keep copies of scripts in `site-packages` which accounts for the difference in success. 

I'm curious if there isn't a difference in the `setuptools` versions between your host and the `virtualenv`. For example, this message is present:

> You are using pip version 9.0.1, however version 19.0.1 is available.

which shows your `virtualenv` is using a version of `pip` that is about three major releases behind. That might indicate that your installed version of `virtualenv` is well out of date as it ships with an embedded version of both `pip` and `setuptools`. You might experiment with this by either upgrading `virtualenv` or adding a newer version of `setuptools` to your `requirements.txt` to see if it starts matching the build strategy it uses when you run `setup.py build` directly.

**Comment by gbtimmon** [comment]
Updating versions didn't seem to help: 

```
(root@tassVM)(/tmp/Issue)#pip list | grep -E '(virtualenv|venvctrl|rpmvenv|setuptools|pip)'
DEPRECATION: Python 2.7 will reach the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 won't be maintained after that date. A future version of pip will drop support for Python 2.7.
pip                              19.0.1
rpmvenv                          0.20.0
setuptools                       40.7.1
setuptools-svn                   1.2
venvctrl                         0.2.1
wheel                            0.32.3
```
Same error as above minus the pip warning message. 

----------------------------------------------------------------------------

I understand your hesitation to avoid supporting EGG formats, they seem to cause more headaches then anything. I may put together a PR for it since I've walked though much of the research for a fix already here, though no promises with my current workload. 

 Its unfortunate that `./setup.py install` defaults to EGG, even when though wheel is available on this machine. A similar `pip install .` will use the wheel if wheel is available, and falls back to ./setup.py install only if no `wheel` package is installed.  With that in mind I thought a better way to go about this is to avoid ./setup.py install. So I try:  `echo '.' >> ./requirements.txt` and setting `'require_setup_py': false`. Seems to do the trick, since I'm using pip and a not completely outdated 'EGG' format now...

I wonder if this should be the default?   Something along the lines of `require_setup_py`: false, and change the pip install statement in the spec to `python install -r requirements.txt .`?

**Comment by kevinconway** [comment]
I feel like there's something we're missing about the environment because `pip install` for a local directory opens a subshell in order to call `setup.py install`. Granted, it wraps the original file with a shim that enforces `setuptools` usage but it's odd that it would result in something different than a `setup.py install` call. For the sake of asking, is the project using `distutils` rather than `setuptools` for the `setup.py`?

As to the method you've found that works, it would be a tough sell to make `pip` the _default_ install method for the project even if that is the better option for modern Python. I try to be overly cautious about changes to defaults in an effort to prevent breaking builds. I'd be accepting, though, of a new optional flag that enables that install method and let folks migrate as needed.

Alternatively, the `requirements` argument of the `python_venv` section is a list which would enable you to have a separate `requirements.txt` type of file that includes the `.` value that is only used during builds. That would enable you to preserve your main `requirements.txt` but swap in the the local install when generating an RPM. I'd like to find a way for this tool to work without that if possible, though.

I'm glad you have a few options for working around the problem. I'm also sorry that we haven't figured out the root cause.

**Comment by gbtimmon** [comment]
I think it all comes down to the venvctrl not supporting the EGG format correctly. The workaround I had hinges on `pip` choosing to build a wheel to install. I double checked that i was using setuptools and I was. Interestingly changing to distutils.core actually fixes the issue as well, as long as you don't mind losing the setuptools features.

./setup.py install with distutils seems to avoid bdist_egg and instead does the following
```
(root@myVM)(/tmp/Issue_distutils)#./setup.py install | grep running
running install
running build
running build_py
running build_scripts
running install_lib
running install_scripts
running install_data
running install_egg_info
```

`./pip install . -v` seems to run install but in a special way 
```
(root@myVM)(/tmp/Issue)# pip install -v .
[...]
Running setup.py install for MyModule ...     Running command /usr/bin/python2 -u -c "import setuptools, tokenize;__file__='/tmp/pip-req-build-cawxD7/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-record-Sdlkvy/install-record.txt --single-version-externally-managed --compile
[...]
```

Summarizing... 
```
(root@myVM)(/tmp/Issue)#./setup.py install --record /tmp/pip-record-Sdlkvy/install-record.txt --single-version-externally-managed --compile | grep running
running install
running build
running build_py
running egg_info
running build_scripts
running install_lib
running install_data
running install_egg_info
running install_scripts
```

And the broken way:
```
(root@tassVM)(/tmp/Issue)#./setup.py install | grep running
running install
running bdist_egg
running egg_info
running install_lib
running build_py
running install_data
running install_scripts
running build_scripts
```

So long story short I think you were right the first time. The only issue is venvctrl doesn't support the bdist_egg format gracefully. The work around's I have are just avoiding that format in various ways. I'd retract my suggestion on the changing of default -- venvctrl just needs to find shebangs in the bdist_egg format correctly.
